src/Language/PureScript/Linter.hs

```diff
@@ -129,3 +130,158 @@ lint (Module _ _ mn ds _) = censor (addHint (ErrorInModule mn)) $ mapM_ lintDecl
+  thisModuleRef _ = False
+
+  rebindable :: S.Set Ident
+  rebindable = S.fromList [ Ident C.bind, Ident C.discard ]
+
+  getDeclIdent :: Declaration -> Maybe Ident
+  getDeclIdent = getIdentName <=< declName
+
+  lintDeclaration :: Declaration -> m (S.Set Ident)
+  lintDeclaration declToLint = do
+    let (vars, errs) = goDecl declToLint
+    tell errs
+    pure vars
+    where
+
+    goDecl :: Declaration -> (S.Set Ident, MultipleErrors)
+    goDecl d@(ValueDeclaration vd) =
+        let allExprs = concatMap unguard $ valdeclExpression vd
+            bindNewNames = S.fromList (concatMap binderNames $ valdeclBinders vd)
+            ss = declSourceSpan d
+            (vars, errs) = removeAndWarn ss bindNewNames $ combineAll $ map (go ss) allExprs
+            errs' = addHint (ErrorInValueDeclaration $ valdeclIdent vd) errs
+        in
+          (vars, errs')
+
+    goDecl (TypeInstanceDeclaration _ _ _ _ _ _ _ (ExplicitInstance decls)) = combineAll $ map goDecl decls
+    goDecl _ = mempty
+
+    go :: SourceSpan -> Expr -> (S.Set Ident, MultipleErrors)
+    go _ (Var _ (Qualified Nothing v)) = (S.singleton v, mempty)
+    go _ (Var _ _) = (S.empty, mempty)
+
+    go ss (Let _ ds e) =
+      let letNames = S.fromList $ concatMap declIdents ds
+      in removeAndWarn ss letNames $ combineAll (go ss e : map underDecl ds)
+    go ss (Abs binder v1) =
+      let newNames = S.fromList (binderNames binder)
+      in
+      removeAndWarn ss newNames $ go ss v1
+
+    go ss (UnaryMinus _ v1) = go ss v1
+    go ss (BinaryNoParens v0 v1 v2) = go ss v0 `combine` go ss v1 `combine` go ss v2
+    go ss (Parens v1) = go ss v1
+    go ss (TypeClassDictionaryConstructorApp _ v1) = go ss v1
+    go ss (Accessor _ v1) = go ss v1
+
+    go ss (ObjectUpdate obj vs) = combineAll (go ss obj : map (go ss . snd) vs)
+    go ss (ObjectUpdateNested obj vs) = go ss obj `combine` goTree vs
+      where
+        goTree (PathTree tree) = combineAll $ map (goNode . snd) (runAssocList tree)
+        goNode (Leaf val) = go ss val
+        goNode (Branch val) = goTree val
+
+    go ss (App v1 v2) = go ss v1 `combine` go ss v2
+    go ss (Unused v) = go ss v
+    go ss (IfThenElse v1 v2 v3) = go ss v1 `combine` go ss v2 `combine` go ss v3
+    go ss (Case vs alts) =
+      let f (CaseAlternative binders gexprs) =
+            let bindNewNames = S.fromList (concatMap binderNames binders)
+                allExprs = concatMap unguard gexprs
+            in
+                removeAndWarn ss bindNewNames $ combineAll $ map (go ss) allExprs
+      in
+      combineAll $ map (go ss) vs ++ map f alts
+
+    go ss (TypedValue _ v1 _) = go ss v1
+    go ss (Do _ es) = doElts ss es Nothing
+    go ss (Ado _ es v1) = doElts ss es (Just v1)
+
+    go ss (Literal _ (ArrayLiteral es)) = combineAll $ map (go ss) es
+    go ss (Literal _ (ObjectLiteral oo)) = combineAll $ map (go ss . snd) oo
+
+    go _ (PositionedValue ss' _ v1) = go ss' v1
+
+    go _ _ = (mempty, mempty)
```

Comment:
I'd quite like to avoid catch-all patterns here, as I think it will be quite easy to forget to update this function if we do add new cases to `Expr` in the future. Would you mind matching every constructor here?